zoom-in-timeseries-drill: do not include :hour and :minute for DATE columns

appleby opened the pull request:
Closes https://github.com/metabase/metabase/issues/39366

### Description

When the user attempts to drill through a DATE column, we should not return `:hour` or `:minute` drill throughs from `available-drill-thrus`. Previously we did, which resulted in an error being displayed when the user attempted to select them.

Modify `next-breakout-unit` in zoom_in_timeseries.cljc to not return :hour or :minute units for :type/Date columns, which prevents these from being returned by `available-drill-thrus`, which prevents them appearing in UI context menus.

### How to verify

Describe the steps to verify that the changes are working as expected.

1. Databases -> Sample Database -> People
2. Summarize -> Count -> Done
3. Breakout by -> Time -> BIRTH_DATE
4. Select "All Time" by "Day" at the bottom
5. Click on a datapoint in the graph
6. Verify that "See this day by hour" does *not* appear in the context menu

You can also do a similar process for a pivot table visualization, and probably 10 other ways I'm not aware of 😬.

### Demo

#### before

<img width="1323" alt="zoom-in-timeseries-before" src="https://github.com/user-attachments/assets/8f7db905-1d57-4079-9cee-33f6ee9d1086">

<img width="1322" alt="before-error" src="https://github.com/user-attachments/assets/4010cf58-5afc-4c0c-92cd-31cd7c822d30">


#### after

<img width="1325" alt="zoom-in-timeseries-after" src="https://github.com/user-attachments/assets/d1eec540-2e77-4a7f-82b6-dfb670fbbbbb">


### Checklist

- [x] Tests have been added/updated to cover changes in this PR

appleby committed:
zoom-in-timeseries-drill: do not include :hour and :minute for DATE columns

When the user attempts to drill through a DATE column, we should not return :hour or :minute drill throughs from
available-drill-thrus. Previously we did, which resulted in an error being displayed when the user attempted to select
them.

Modify next-breakout-unit in zoom_in_timeseries.cljc to not return :hour or :minute units for :type/Date columns,
which prevents these from being returned by available-drill-thrus, which prevents them appearing in UI context menus.

ranquild commented on src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:
```diff
@@ -48,13 +50,26 @@
-;;; TODO -- we shouldn't include hour and minute for `:type/Date` columns.
-(def ^:private valid-current-units
-  [:year :quarter :month :week :day :hour :minute])
+(mu/defn- valid-current-units-for-schema-type :- [:vector :keyword]
```
We could use `lib.temporal-bucket/available-temporal-buckets` here. It returns all available options for per column type; but it seems that we need truncation units only. We could use `lib.schema.temporal-bucketing/datetime-truncation-units` for this. Maybe something like:
```
(into []
  (filter (fn [option]
            (->> (lib.temporal-bucket/raw-temporal-bucket option)
                 (contains? lib.schema.temporal-bucketing/datetime-extraction-units))))
  (lib.temporal-bucket/available-temporal-buckets query stage-number column))
```

As a bonus, it:
1. Actually handles `effective-type` correctly
2. Handles custom expressions for which we don't allow temporal buckets in breakouts

appleby replied:
Hi @ranquild, thanks for the review!

I'm glad you mention `lib.temporal-bucket` and `:effective-type` as these were two things I was going to ask about specifically.

In particular, I don't use `available-temporal-buckets` because it is checking `:effictive-type` which turns out to be `:type/Integer` when bucketing by `:year`.

``` clojure
(-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
    (lib/breakout (-> (meta/field-metadata :orders :created-at)
                      (lib/with-temporal-bucket :year)))
    :stages
    first
    :breakout)
==>
[[:field
  {:lib/uuid "d2588602-8018-42d2-92f6-83fb8e7e392b",
   :base-type :type/DateTimeWithLocalTZ,
   :effective-type :type/Integer,
   :temporal-unit :year,
   :metabase.lib.field/original-effective-type :type/DateTimeWithLocalTZ}
  57606]]
```

And, consequently:

``` clojure
(let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
                (lib/aggregate (lib/count))
                (lib/breakout (-> (meta/field-metadata :orders :created-at)
                                  (lib/with-temporal-bucket :year))))
      option {:lib/type :metadata/column,
              :base-type :type/DateTimeWithLocalTZ,
              :effective-type :type/Integer}]
  (lib.temporal-bucket/available-temporal-buckets query -1 option))

==>

[]
```

See, e.g. the comment at the top of

``` clojure
(defmethod lib.temporal-bucket/with-temporal-bucket-method :field
  [[_tag options id-or-name] unit]
  ;; if `unit` is an extraction unit like `:month-of-year`, then the `:effective-type` of the ref changes to
  ;; `:type/Integer` (month of year returns an int). We need to record the ORIGINAL effective type somewhere in case
  ;; we need to refer back to it, e.g. to see what temporal buckets are available if we want to change the unit, or if
  ;; we want to remove it later. We will record this with the key `::original-effective-type`. Note that changing the
  ;; unit multiple times should keep the original first value of `::original-effective-type`.
  (if unit
    (let [extraction-unit?        (contains? lib.schema.temporal-bucketing/datetime-extraction-units unit)
          original-effective-type ((some-fn ::original-effective-type :effective-type :base-type) options)
		  ...
```

The reason being that `lib.schema.temporal-bucketing/datetime-extraction-units` contains `:year`.

That is why I check for `original-effective-type` and `base-type` rather than `effective-type`, but I wanted to ask about that as I was unsure whether that was the right approach...

Also thanks for the tip about filtering `available-temporal-buckets`. I had seen `lib.schema.temporal-bucketing/datetime-truncation-units` but noticed that it was (a) unordered (there is an `ordered-date-truncation-units` but no corresponding `ordered-datetime-truncation-units`, though I guess it would be easy to add) and (b) contains "extra" units like `:second` and `:millisecond` that the old code here didn't include. But I see now that filtering `available-temporal-buckets` by `datetime-truncation-units` seems to be exactly what we'd want, except for the case of `:year` buckets mentioned above.

ranquild replied:
it's usually incorrect to check for either `base-type` or `effective-type` manually. There are many cases to be aware of, like coercion strategies, buckets changing the type etc.

ranquild replied:
are you sure that the behavior you want to achieve with `:year` is correct?

ranquild replied:
@bshepherdson please take a look

bshepherdson replied:
I think the right way to do this is using `available-temporal-buckets-method` on the breakout ref. That refers to the type of the original field, ignoring any bucketing, which should be `Date`, `DateTime`, or `Time`, and return the right set.

So I'd approach it like this:
- `lib.temporal-bucket/available-temporal-buckets` to get the list for that column
- create `lib.schema.temporal-bucketing/ordered-datetime-truncation-units` since it doesn't exist yet
- filter the `ordered-datetime-truncation-units` to only those which are `available`
- the next one in that list is the one!

appleby replied:
@bshepherdson how do I get the breakout ref? I'm guessing it's something other than the thing I'm selecting in my example above, but I didn't see anything in the query or context that is passed in that looks like a reference to a breakout with the original field's `effective-type`.

In other words, not this, I assume:

```clojure
(-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
    (lib/breakout (-> (meta/field-metadata :orders :created-at)
                      (lib/with-temporal-bucket :year)))
    :stages
    first
    :breakout)
==>
[[:field
  {:lib/uuid "d2588602-8018-42d2-92f6-83fb8e7e392b",
   :base-type :type/DateTimeWithLocalTZ,
   :effective-type :type/Integer,
   :temporal-unit :year,
   :metabase.lib.field/original-effective-type :type/DateTimeWithLocalTZ}
  57606]]
```

Or can I use that uuid to look it up?

bshepherdson replied:
That's the correct ref. The right way to access it is with `(first (lib.breakout/breakouts query -1))`, though. Disregard its `:effective-type`; `available-temporal-buckets` should do the right thing anyway.

(It looks up the original input field rather than relying on the breakout directly, see the `(defmethod available-temporal-buckets-method :field ...)` in `lib.field`.)

bshepherdson replied:
For context, this nonsense is the kind of thing I mean when I said in the retro this morning that refs suck and are the source of much of our pain, bugs and perf waste.

In the prototype I mentioned, breakouts are themselves columns and contain a reference (the ID of another entity in the "database"; which is to say a real, tangible reference rather than a symbolic label!) so you can navigate to the input column with a `get`: `(:mbql.breakout/column a-breakout)` and done.

appleby replied:
Yep, that works! I will rework to use this method instead. Thanks!

```clojure
(let [query (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
                (lib/aggregate (lib/count))
                (lib/breakout (-> (meta/field-metadata :orders :created-at)
                                  (lib/with-temporal-bucket :year))))
      breakout (first (lib.breakout/breakouts query -1))]
  (into []
        (comp 
         (map lib.temporal-bucket/raw-temporal-bucket)
         (filter #(contains? lib.schema.temporal-bucketing/datetime-truncation-units %)))
        (lib.temporal-bucket/available-temporal-buckets query -1 breakout)))

==>

[:minute :hour :day :week :month :quarter :year]
```

appleby commented:
AFAICT, the test failures look unrelated to this PR. The failing e2e test appears to be flake. The failing bigquery driver test has a name that looks like it could possibly be something broken by this PR, but the actual error appears to just be a read timeout syncing metadata, so maybe not? Not sure what the protocol is for re-running tests when they are suspected flakes...

```
ERROR in metabase.query-processor-test.date-bucketing-test/count-of-grouping-test-2 (get_or_create.clj:145)

:bigquery-cloud-sdk
 4 checkins per hour dataset group by hour
filter by hour = [:relative-datetime :current]
expected: (= 4 (apply count-of-grouping checkins:4-per-hour :hour args))
  actual: 
                 clojure.lang.ExceptionInfo: Failed to sync test database "interval_900": Read timed out
```

appleby committed:
Use lib.temporal-bucket/available-temporal-buckets

Per review feedback, use lib.temporal-bucket/available-temporal-buckets to get the valid-current-units in
metabase.lib.drill-thru.zoom-in-timeseries.

appleby committed:
Do not construct zipmap just to lookup the next unit

Rather than constructing a zipmap from `valid-current-units', which requires iterating over the seq twice, just
traverse the seq once and return the next item in the seq (if any).

Constructing a zipmap made sense in the old code because it was constructed once from static data and used many times
for lookup. In the new version, the vec of `valid-current-units` is dynamic and dependent on the type of the breakout.

appleby commented on src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:
```diff
@@ -69,11 +70,13 @@
+(mu/defn- next-breakout-unit :- [:maybe ::lib.schema.drill-thru/drill-thru.zoom-in.timeseries.next-unit]
+  [query :- ::lib.schema/query
+   stage :- :int
+   field :- :mbql.clause/field]
```
Is `:mbql.clause/field` the right type here? I'm assuming that the `(lib.util/clause-of-type? breakout :field)` in `matching-breakout-dimension` means we're guaranteed to get a `:mbql.clause/field` here.

ranquild replied:
I'd use `column` everywehre

bshepherdson replied:
Yes, I think that's correct here.

appleby commented:
@ranquild @bshepherdson

This is ready for re-review whenever you get a chance.

ranquild reviewed (commented):
Looks good! I'll let @bshepherdson approve. I think `field :- :mbql.clause/field` might be wrong.

bshepherdson commented on src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:
```diff
@@ -41,20 +43,19 @@
+  [query :- ::lib.schema/query
+   stage :- :int
+   field :- :mbql.clause/field]
+  (into [] (reverse (eduction (map lib.temporal-bucket/raw-temporal-bucket)
+                              (filter #(contains? lib.schema.temporal-bucketing/datetime-truncation-units %))
+                              (lib.temporal-bucket/available-temporal-buckets query stage field)))))
```
This is a lot harder to read than a direct `->>` or `(reverse (into [] (comp (map ...) (filter ...)) ...))`.

Also, note that sets can be used as functions, so there's no need for `contains?` here; just `(filter lib.schema.temporal-bucketing/datetime-truncation-units)`.

You could write this as
```clojure
(reverse (into [] (comp (map lib.temporal-bucket/raw-temporal-bucket)
                        (filter lib.schema.temporal-bucketing/datetime-truncation-units))
                  (lib.temporal-bucket/available-temporal-buckets query stage field)))
```

but I honestly think it'll be better just to use `->>` naively. This list is too short to worry about any overhead.

I'll also point out in passing that `(comp f g)` applies `g`, then `f`, ie. it's right-to-left. **But** transducers are functions that wrap an abstract pipeline with a new, pre-order transformation. So `(comp (map f) (filter g))` produces a transducer that maps `f` and then filters with `g`.

```clojure
(->> (lib.temporal-bucket/available-temporal-buckets query stage field)
     (map lib.temporal-bucket/raw-temporal-bucket)
     (filter lib.schema.temporal-bucketing/datetime-truncation-units)
     reverse)
```

appleby replied:
Awesome, thanks for the tip. As you can tell I'm still struggling with idiomatic Clojure, so these tips are appreciated. Will update the PR with these suggestions.

appleby replied:
Fixed

bshepherdson commented on src/metabase/lib/drill_thru/zoom_in_timeseries.cljc:
```diff
@@ -69,11 +70,13 @@
+   stage :- :int
+   field :- :mbql.clause/field]
+  (when-let [current-unit (lib.temporal-bucket/raw-temporal-bucket field)]
+    (second (drop-while #(not= % current-unit)
+                        (valid-current-units query stage field)))))
```
```suggestion
    (->> (valid-current-units query stage field)
         (drop-while #(not= % current-unit))
         second)))
```

appleby committed:
PR suggestion: use threading macro rather than nested sexps

Co-authored-by: Braden Shepherdson <braden@metabase.com>

appleby committed:
PR suggestion: use threading macro rather than transducer / eduction

appleby commented:
Rebased onto master and updated the label from `no-backport` to `backport`

camsaul commented on test/metabase/lib/drill_thru/zoom_in_timeseries_test.cljc:
```diff
@@ -134,6 +136,81 @@
+    :custom-row   {"count" 77, "sum" 1, "max" nil, "CREATED_AT" "2022-12-01"}
+    :expected     {:type :drill-thru/zoom-in.timeseries}}))
+
+(deftest ^:parallel returns-zoom-in-timeseries-test-4
```
A lot of our tests end in `test-n` or whatever as a byproduct of breaking out giant tests into smaller ones, it's even better to give different tests more meaningful names if you can. Something like `returns-zoom-in-timeseries-count-date-by-month-test` or something and then `-sum-date-by-month-test` for the next one is a little bit better than just giving every test a number .

I would really consider combining these all into one test tho, maybe something like

```clj
(deftest ^:parallel returns-zoom-in-timeseries-date-column-test
  (testing "zoom-in.timeseries should be returned for Date dimension (#33811, #39366)"
    (doseq [aggregation ["count" "sum" "max"]
            bucketing   [:month :day]]
      (testing (lib.util/format "aggregation = %s, bucketing = %s" aggregation bucketing)
        (lib.drill-thru.tu/test-returns-drill
         (zoom-in-timeseries-drill-for-orders-created-at
          metadata-provider-with-orders-created-at-as-date
          aggregation
          bucketing))))))
```

otherwise it gets really repetitive and noisy to write out 6 versions of the same test that only differ by one or two things

appleby replied:
Fixed in latest commit

camsaul reviewed (approved):
Well overall this LGTM but I would consider combining all those very similar tests into one single test with `doseq` to make things a little more concise before merging

appleby commented:
> Well overall this LGTM but I would consider combining all those very similar tests into one single test with `doseq` to make things a little more concise before merging

That's actually how I implemented it in a first pass, then noticed the existing tests for DateTime were separate tests and copied that, thinking maybe we prefer to break them out to increase parallelism or else just prefer to have separate test cases to make it easier to identify which test failed.

Will update before merging.

appleby committed:
PR suggestion: reduce boilerplate in zoom-in-timeseries tests

Reduce code dup by looping over aggregation and bucketing options rather than having a separate deftest for each.
